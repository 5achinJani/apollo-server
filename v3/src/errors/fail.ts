import { Class, Instance, isTemplateInvocation } from '../utilities/types'

export function fail
  (code: string): FailureMode<[]>
export function fail<P extends any[], M extends FailureMode<P, any>>
  (code: string, mode: M): M
export function fail<E extends Class<Error>>
  (code: string, Base: E): FailureMode<ConstructorParameters<E>, Instance<E>>
export function fail<E extends ErrorWithStaticCode>
  (Base: E): FailureMode<ConstructorParameters<E>, Instance<E>>

export function fail<E extends ErrorWithStaticCode>(
  codeOrBase: E | string,
  ...rest: any[]
): any {
  let code: string
  let Base: Class<Error>
  if (typeof codeOrBase === 'string') {
    code = codeOrBase
    Base = rest[1] || Error
  } else {
    Base = rest[0]
    code = codeOrBase.code
  }

  return failureMode(Base, code)
}


export type Message<P extends any[]> = [string] | [(...params: P) => string] | [TemplateStringsArray, ...any[]]

export type Fail<P extends any[], Base extends Error> = Base & {
  readonly code: string
  readonly params: P
}

export interface FailureMode<P extends any[] = [object?], Base extends Error = Error> {
  (...params: P): Failure<P, Base>
  readonly code: string
  class: Class<Fail<P, Base>> & (new (fromFailure: Failure<P, Base>) => Fail<P, Base>)
  message: <MoreProps=undefined>(...msg: Message<[MoreProps]>) =>
    P extends []
      ? MoreProps extends object
        ? FailureMode<[MoreProps], Base>
        : FailureMode<P, Base>
      :
    P extends [infer Props]
      ? MoreProps extends object
        ? FailureMode<[Props & MoreProps], Base>
        : FailureMode<P, Base>
      :
      FailureMode<P, Base>
}

export interface Failure<P extends any[], Base extends Error> {
  create(): Fail<P, Base>
  readonly messages: Message<P>[]
}

export class Failure<P extends any[], Base extends Error> {
  constructor(
    public readonly mode: FailureMode<P, Base>,
    public readonly params: P,
    public readonly messages: Message<P>[] = []) {}

  create(): Fail<P, Base> {
    return new this.mode.class(this)
  }
}

export interface ErrorWithStaticCode extends Class<Error> {
  readonly code: string
}

function formatMsg<P extends any[]>(message: Message<P>, params: P) {
  if (!message.length) return null
  const [format] = message
  console.log('format:', format)
  if (typeof format === 'function') {
    console.log(format, params)
    return format(...params)
  }
  if (isTemplateInvocation(message)) return String.raw(...message)
  if (typeof format === 'string') return format
  return null
}

const FAILURE_CODE = Symbol('Failure code')

function failureMode<B extends Class<Error>>(Base: B, code: string, ...messages: Message<ConstructorParameters<B>>[]): FailureMode<ConstructorParameters<B>, Instance<B>> {
  const FailureClass = ((Base as any)[FAILURE_CODE] === code) ? Base :
    class FailureClass extends Base {
      static readonly [FAILURE_CODE] = code
      static readonly code = code
      public readonly failure: Failure<ConstructorParameters<B>, Instance<B>>

      constructor(...args: any[]) {
        const [failure] = args as [Failure<ConstructorParameters<B>, Instance<B>>]
        super(...failure.params)
        this.failure = failure
      }

      get message() {
        if (!super.message && !this.failure.messages.length) return code
        return [
          super.message,
          ...this.failure.messages.map(m => formatMsg(m, this.failure.params))
        ].filter(Boolean).join('\n\n')
      }
    }

  function FailureMode(...args: ConstructorParameters<B>) {
    return new Failure(FailureMode as any, args, messages)
  }

  Object.defineProperties(FailureClass, {
    name: { value: code },
  })

  Object.defineProperties(FailureMode, {
    code: {value: code},
    message: {
      value(...message: Message<ConstructorParameters<B>>) {
        return failureMode(FailureClass, code, ...messages, message)
      }
    }
  })

  FailureMode.prototype = Object.create(FailureClass.prototype)
  FailureMode.prototype.constructor = FailureMode

  return FailureMode as any
}
