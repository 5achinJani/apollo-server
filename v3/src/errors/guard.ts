import { fail, FailureMode, Failure } from "./fail";
import { Class } from "../utilities/types";

const E_GUARD = fail('GUARD')
const E_CHECK = fail('CHECK')

export type Comparison = '<' | '>' | '==' | '===' | '!=' | '!==' | '<=' | '>='
const COMPARE = {
  '<': (lhs: any, rhs: any) => lhs < rhs,
  '>': (lhs: any, rhs: any) => lhs > rhs,
  '==': (lhs: any, rhs: any) => lhs == rhs,
  '===': (lhs: any, rhs: any) => lhs === rhs,
  '!=': (lhs: any, rhs: any) => lhs != rhs,
  '!==': (lhs: any, rhs: any) => lhs !== rhs,
  '<=': (lhs: any, rhs: any) => lhs <= rhs,
  '>=': (lhs: any, rhs: any) => lhs >= rhs,
}

export type ToThrow = FailureMode | Failure<any>

function use(toThrow: ToThrow) {
  return 'create' in toThrow ? toThrow : toThrow()
}

export function guard(condition: any, failure: ToThrow = E_GUARD): asserts condition {
  if (!condition) {
    throw use(failure).msg `Guard failed`.create()
  }
}

export function check(lhs: any, op: Comparison, rhs: any, failure: ToThrow = E_CHECK): boolean {
  if (op in COMPARE && COMPARE[op](lhs, rhs)) return true
  throw use(failure).msg `Check failed: ${lhs} ${op} ${rhs}`.create()
}

check.instance = <R>(lhs: any, cls: Class<R>, failure: ToThrow = E_CHECK): asserts lhs is R => {
  if (!(lhs instanceof cls)) {
    throw use(failure).msg `Check failed: ${lhs} instanceof ${cls.name}`.create()
  }
}

check.exists = (o: any, failure: ToThrow = E_CHECK): o is Exclude<any, null | undefined> => {
  if (o == null) {
    throw use(failure).msg `Check failed: ${o} exists`.create()
  }
  return true
}
