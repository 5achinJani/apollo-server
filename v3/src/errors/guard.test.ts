import { guard, Comparison } from "./guard";
import { fail } from "./fail";

describe("guard", () => {
  it("asserts conditions", () => {
    const a = "hello";
    expect(() => {
      guard(a !== a);
    }).toThrowErrorMatchingInlineSnapshot(`"Assertion failed"`);
  });

  it("passes when condition ok", () => {
    const a = "hello";
    expect(() => {
      guard(a === a);
    }).not.toThrow();
  });

  it(".instance asserts instanceof", () => {
    expect(() => {
      guard.instance(2, String);
    }).toThrowErrorMatchingInlineSnapshot(
      `"Assertion failed: 2 instanceof String"`
    );
  });

  it(".instance passes when ok", () => {
    expect(() => {
      guard.instance(new String("hello"), String);
    }).not.toThrow();
  });

  const comparisons: Comparison[] = ['<', '>', '==', '===', '!=', '!==', '<=', '>=']
  const obj = {}

  it.each([...function*() {
    const obj = {}
    const examples = new Map<any, any>()
    examples.set(0, [-1, 1, 0, 100, -Infinity, Infinity])
    examples.set('hello', ['hello', '', obj])
    examples.set(obj, [{}, obj, 'a string'])
    for (const lhs of examples.keys()) {
      for (const rhs of examples.get(lhs)) {
        for (const op of comparisons) {
          yield [lhs, op, rhs]
        }
      }
    }
  }()])("compares %s %s %s", ((lhs, op, rhs) => {
    const func = `((l, r) => l ${op} r)`
    const run = expect(() => {
      guard(lhs, op, rhs)
    })
    if (eval(func)(lhs, rhs)) {
      run.not.toThrow()
    } else {
      run.toThrowError(`Assertion failed: ${lhs} ${op} ${rhs}`)
    }
  }));

  it('compares objects', () => {
    guard(obj, '==', obj)
  })

  const E_TOO_FAST = fail("TOO_FAST").msg(
    (props: { maxRate: number, rate: number, client: string }) =>
      `Client ${props.client} exceeded ${props.maxRate} with rate=${props.rate}`
  );
  it("optionally, takes a failure to guard with", () => {
    const maxRate = 1000;
    const rate = 9999;
    const client = "a-client";
    expect(() => {
      guard(rate, "<=", maxRate, E_TOO_FAST({ maxRate, rate, client }));
    }).toThrowErrorMatchingInlineSnapshot(`
"Client a-client exceeded 1000 with rate=9999

Assertion failed: 9999 <= 1000"
`);
  });
});
