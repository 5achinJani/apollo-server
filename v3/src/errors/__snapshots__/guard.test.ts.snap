// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guard performs comparisons 1`] = `"Assertion failed: hello === world"`;
